File: plugins/sap-architecture/skills/sap-architecture/scripts/extract_icon.py
#!/usr/bin/env python3
"""Extract a ready-to-paste <mxCell> for a named SAP BTP service icon.

Usage:
  extract_icon.py <service-name> [--x X --y Y --w W --h H --id ID --parent P --label "text"]
  extract_icon.py --list

Search is fuzzy (case-insensitive substring + slug alias).
Emits the <mxCell ...><mxGeometry .../></mxCell> to stdout.
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from xml.sax.saxutils import escape

HERE = Path(__file__).resolve().parent
INDEX = HERE.parent / "assets" / "icon-index.json"
ASSET_INDEX = HERE.parent / "assets" / "asset-index.json"

try:
    from extract_asset import emit_asset as emit_general_asset
    from extract_asset import find_asset as find_general_asset
except ImportError:  # pragma: no cover - fallback for copied standalone script use
    emit_general_asset = None
    find_general_asset = None


def load_index() -> dict:
    if not INDEX.exists():
        print(f"icon index not found at {INDEX}; run build_icon_index.py first", file=sys.stderr)
        sys.exit(1)
    return json.loads(INDEX.read_text(encoding="utf-8"))


def load_asset_index() -> dict | None:
    if not ASSET_INDEX.exists() or emit_general_asset is None or find_general_asset is None:
        return None
    return json.loads(ASSET_INDEX.read_text(encoding="utf-8"))


def _normalize(text: str) -> str:
    """Strip HTML entities (&#10;, &amp;), lowercase, collapse non-alphanumeric to single spaces."""
    t = text.lower()
    # decode common HTML newline / ampersand entities
    t = re.sub(r"&#?\w+;", " ", t)
    # kill the -10- slug artifact from earlier encoding
    t = re.sub(r"(^|-)10(-|$)", r"\1 \2", t)
    return re.sub(r"[^a-z0-9]+", " ", t).strip()


def _tokens(text: str) -> set[str]:
    return {tok for tok in _normalize(text).split() if tok}


# Hand-curated aliases for common SAP service abbreviations / nicknames.
# These map a user-typed query → a slug substring that uniquely identifies the icon.
COMMON_ALIASES: dict[str, str] = {
    "xsuaa": "sap-authorization-10-and-10-trust-management-service",
    "uaa": "sap-authorization-10-and-10-trust-management-service",
    "auth trust": "sap-authorization-10-and-10-trust-management-service",
    "ias": "identity-10-authentication",
    "ips": "identity-provisioning",
    "btp cockpit": "sap-btp-cockpit",
    "hana": "sap-hana-cloud",
    "hana cloud": "sap-hana-cloud",
    "cf": "cloud-10-foundry-runtime",
    "cloud foundry": "cloud-10-foundry-runtime",
    "cap": "cloud-application-programming",
    "destination": "sap-destination-10-service",
    "cc": "cloud-10-connector",
    "connector": "cloud-10-connector",
    "abap env": "abap-environment",
    "abap environment": "abap-environment",
    "build apps": "sap-build-apps",
    "build process": "sap-build-process-automation",
    "build code": "sap-build-code",
    "joule": "joule-studio",
    "task center": "sap-task-center",
    "cpi": "cloud-10-integration",
    "integration suite": "integration-suite",
}


def find(index: dict, query: str) -> tuple[str, dict] | None:
    q = query.lower().strip()
    qslug = re.sub(r"[^a-z0-9]+", "-", q).strip("-")
    # curated alias table
    if q in COMMON_ALIASES:
        slug = COMMON_ALIASES[q]
        if slug in index:
            return slug, index[slug]
    # exact slug
    if qslug in index:
        return qslug, index[qslug]
    # alias match
    for slug, entry in index.items():
        if qslug in entry["aliases"] or qslug == entry["display"].lower():
            return slug, entry

    qtokens = _tokens(q)
    if not qtokens:
        return None

    # token-subset match: every query token appears in (display | slug | aliases)
    token_candidates: list[tuple[str, dict]] = []
    for slug, entry in index.items():
        pool = _tokens(slug) | _tokens(entry["display"]) | {a for alias in entry["aliases"] for a in _tokens(alias)}
        if qtokens <= pool:
            token_candidates.append((slug, entry))
    if len(token_candidates) == 1:
        return token_candidates[0]
    if len(token_candidates) > 1:
        # prefer the candidate whose normalized display is shortest (tightest match)
        token_candidates.sort(key=lambda se: len(_normalize(se[1]["display"])))
        # if the top candidate is strictly shorter than the runner-up, take it
        top = _normalize(token_candidates[0][1]["display"])
        second = _normalize(token_candidates[1][1]["display"])
        if len(top) < len(second):
            return token_candidates[0]
        print(f"ambiguous '{query}' — {len(token_candidates)} matches:", file=sys.stderr)
        for s, e in token_candidates[:10]:
            print(f"  {s}  —  {e['display']}", file=sys.stderr)
        sys.exit(2)

    # substring fallback
    candidates = [(s, e) for s, e in index.items() if q in e["display"].lower() or qslug in s]
    if len(candidates) == 1:
        return candidates[0]
    if len(candidates) > 1:
        print(f"ambiguous '{query}' — {len(candidates)} matches:", file=sys.stderr)
        for s, e in candidates[:10]:
            print(f"  {s}  —  {e['display']}", file=sys.stderr)
        sys.exit(2)
    return None


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("query", nargs="?")
    ap.add_argument("--list", action="store_true")
    ap.add_argument("--x", type=int, default=0)
    ap.add_argument("--y", type=int, default=0)
    # Defaults match the SAP corpus convention: 32x32 is the dominant icon
    # size (used 224x across the 71 bundled templates), then 48x48 (157x).
    # The previous default 64x80 caused icons to overlap card text — the
    # most common visible bug. Override with --w 48 --h 48 only when the
    # icon is the focal anchor of a zone (e.g. a brand mark beside a
    # top-left zone label).
    ap.add_argument("--w", type=int, default=32)
    ap.add_argument("--h", type=int, default=32)
    ap.add_argument("--id", default="icon1")
    ap.add_argument("--parent", default="1")
    ap.add_argument("--label", default=None, help="Override icon label text")
    args = ap.parse_args()

    asset_index = load_asset_index()
    index = load_index()
    if args.list:
        if asset_index:
            assets = asset_index["assets"]
            for key in sorted(assets):
                entry = assets[key]
                if entry["kind"] == "btp-service-icon":
                    slug = key.split(":", 1)[1]
                    print(f"{slug:60s}  {entry['display']}")
        else:
            for slug in sorted(index):
                print(f"{slug:60s}  {index[slug]['display']}")
        return 0
    if not args.query:
        ap.print_usage(sys.stderr)
        return 2

    if asset_index and find_general_asset and emit_general_asset:
        asset_match = find_general_asset(asset_index, args.query, "btp-service-icon")
        if asset_match:
            slug, asset = asset_match
            print(emit_general_asset(asset, args))
            print(f"# matched: {slug} — {asset['display']}", file=sys.stderr)
            return 0

    match = find(index, args.query)
    if not match:
        print(f"no icon matches '{args.query}'", file=sys.stderr)
        return 1
    slug, entry = match

    style = entry["style"]
    label = args.label if args.label is not None else entry["label"]
    label_xml = escape(label, {'"': "&quot;"})

    # Snap to 10-px grid (quiet fix)
    x = round(args.x / 10) * 10
    y = round(args.y / 10) * 10
    w = round(args.w / 10) * 10
    h = round(args.h / 10) * 10

    cell = (
        f'<mxCell id="{args.id}" value="{label_xml}" style="{style}" '
        f'vertex="1" parent="{args.parent}">'
        f'<mxGeometry x="{x}" y="{y}" width="{w}" height="{h}" as="geometry"/>'
        f"</mxCell>"
    )
    print(cell)
    print(f"# matched: {slug} — {entry['display']}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/tools/extract-icon.js
import { readFileSync } from "node:fs";
import { iconIndexPath } from "../core/paths.js";
import { snap10 } from "../core/round.js";
import { xmlEscapeAttr } from "../core/labels.js";
import { extractAsset, findAsset, loadAssetIndex, normalize as normalizeAsset } from "./extract-asset.js";
import { fail, ok } from "./result.js";

export const COMMON_ALIASES = {
  xsuaa: "sap-authorization-10-and-10-trust-management-service",
  uaa: "sap-authorization-10-and-10-trust-management-service",
  "auth trust": "sap-authorization-10-and-10-trust-management-service",
  ias: "identity-10-authentication",
  ips: "identity-provisioning",
  "btp cockpit": "sap-btp-cockpit",
  hana: "sap-hana-cloud",
  "hana cloud": "sap-hana-cloud",
  cf: "cloud-10-foundry-runtime",
  "cloud foundry": "cloud-10-foundry-runtime",
  cap: "cloud-application-programming",
  destination: "sap-destination-10-service",
  cc: "cloud-10-connector",
  connector: "cloud-10-connector",
  "abap env": "abap-environment",
  "abap environment": "abap-environment",
  "build apps": "sap-build-apps",
  "build process": "sap-build-process-automation",
  "build code": "sap-build-code",
  "build work zone": "sap-build-work-zone-10-standard-edition",
  "sap build work zone": "sap-build-work-zone-10-standard-edition",
  "work zone": "sap-build-work-zone-10-standard-edition",
  joule: "joule-studio",
  "task center": "sap-task-center",
  cpi: "cloud-10-integration",
  "integration suite": "integration-suite"
};

export function loadIconIndex() {
  return JSON.parse(readFileSync(iconIndexPath, "utf8"));
}

function normalize(text = "") {
  let value = String(text).toLowerCase();
  value = value.replace(/&#?\w+;/g, " ");
  value = value.replace(/(^|-)10(-|$)/g, "$1 $2");
  return value.replace(/[^a-z0-9]+/g, " ").trim();
}

function tokens(text = "") {
  return new Set(normalize(text).split(/\s+/).filter(Boolean));
}

function isSubset(needle, haystack) {
  for (const value of needle) {
    if (!haystack.has(value)) return false;
  }
  return true;
}

export function isBackendSystemQuery(query) {
  const normalized = normalize(query);
  const compact = String(query).toLowerCase().replace(/[^a-z0-9]+/g, "");
  return (
    ["sap s 4hana", "sap s 4hana cloud", "s 4hana", "s 4hana cloud"].includes(normalized) ||
    ["s4hana", "saps4hana", "s4hanacloud", "saps4hanacloud"].includes(compact)
  );
}

export function backendSystemGuidance(query) {
  return `'${query}' is a backend system/product label, not a BTP service icon. Use extract_asset.py "sap-s-4hana" --kind sap-brand-name for the label or extract_asset.py "on-premise-sap" --kind generic-icon for a generic backend icon.`;
}

export function findIcon(index, query) {
  if (isBackendSystemQuery(query)) return null;
  const q = String(query).toLowerCase().trim();
  const qslug = q.replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  if (COMMON_ALIASES[q] && index[COMMON_ALIASES[q]]) return [COMMON_ALIASES[q], index[COMMON_ALIASES[q]]];
  if (index[qslug]) return [qslug, index[qslug]];

  for (const [slug, entry] of Object.entries(index)) {
    if ((entry.aliases || []).includes(qslug) || qslug === String(entry.display).toLowerCase()) {
      return [slug, entry];
    }
  }

  const qtokens = tokens(q);
  if (!qtokens.size) return null;
  const tokenCandidates = [];
  for (const [slug, entry] of Object.entries(index)) {
    const pool = new Set([...tokens(slug), ...tokens(entry.display)]);
    for (const alias of entry.aliases || []) {
      for (const token of tokens(alias)) pool.add(token);
    }
    if (isSubset(qtokens, pool)) tokenCandidates.push([slug, entry]);
  }
  if (tokenCandidates.length === 1) return tokenCandidates[0];
  if (tokenCandidates.length > 1) {
    tokenCandidates.sort((a, b) => normalize(a[1].display).length - normalize(b[1].display).length);
    const top = normalize(tokenCandidates[0][1].display);
    const second = normalize(tokenCandidates[1][1].display);
    if (top.length < second.length) return tokenCandidates[0];
    const error = new Error(`ambiguous '${query}' - ${tokenCandidates.length} matches`);
    error.matches = tokenCandidates.slice(0, 10);
    throw error;
  }

  const candidates = Object.entries(index).filter(([slug, entry]) => q && (String(entry.display).toLowerCase().includes(q) || slug.includes(qslug)));
  if (candidates.length === 1) return candidates[0];
  if (candidates.length > 1) {
    const error = new Error(`ambiguous '${query}' - ${candidates.length} matches`);
    error.matches = candidates.slice(0, 10);
    throw error;
  }
  return null;
}

function formatAmbiguous(error) {
  if (!error.matches) return `${error.message}\n`;
  const lines = [`${error.message}:`];
  for (const [slug, entry] of error.matches) {
    lines.push(`  ${slug}  -  ${entry.display}`);
  }
  return `${lines.join("\n")}\n`;
}

export function extractIcon(options) {
  const opts = {
    x: Number(options.x ?? 0),
    y: Number(options.y ?? 0),
    w: Number(options.w ?? 32),
    h: Number(options.h ?? 32),
    id: options.id || "icon1",
    parent: options.parent || "1",
    label: options.label,
    query: options.query,
    list: Boolean(options.list)
  };

  const index = loadIconIndex();
  if (opts.list) {
    let assetIndex = null;
    try {
      assetIndex = loadAssetIndex();
    } catch {
      assetIndex = null;
    }
    if (assetIndex) {
      const lines = [];
      for (const key of Object.keys(assetIndex.assets).sort()) {
        const entry = assetIndex.assets[key];
        if (entry.kind === "btp-service-icon") {
          lines.push(`${key.split(":", 2)[1].padEnd(60)}  ${entry.display}`);
        }
      }
      return ok(`${lines.join("\n")}${lines.length ? "\n" : ""}`);
    }
    const lines = Object.keys(index).sort().map((slug) => `${slug.padEnd(60)}  ${index[slug].display}`);
    return ok(`${lines.join("\n")}${lines.length ? "\n" : ""}`);
  }
  if (!opts.query) return fail(2, "usage: extract-icon <query>\n");
  if (isBackendSystemQuery(opts.query)) return fail(1, `${backendSystemGuidance(opts.query)}\n`);

  const preferIconIndex = Object.hasOwn(COMMON_ALIASES, String(opts.query).toLowerCase().trim());
  if (!preferIconIndex) {
    try {
      const assetIndex = loadAssetIndex();
      const assetMatch = findAsset(assetIndex, opts.query, "btp-service-icon");
      if (assetMatch) {
        return extractAsset({ ...opts, kind: "btp-service-icon", id: opts.id || "icon1" });
      }
    } catch {
      // Fall back to the legacy icon index below.
    }
  }

  let match;
  try {
    match = findIcon(index, opts.query);
  } catch (error) {
    return fail(2, formatAmbiguous(error));
  }
  if (!match) return fail(1, `no icon matches '${opts.query}'\n`);
  const [slug, entry] = match;
  const label = opts.label ?? entry.label;
  const cell = `<mxCell id="${xmlEscapeAttr(opts.id)}" value="${xmlEscapeAttr(label)}" style="${xmlEscapeAttr(entry.style)}" vertex="1" parent="${xmlEscapeAttr(opts.parent)}"><mxGeometry x="${snap10(opts.x)}" y="${snap10(opts.y)}" width="${snap10(opts.w)}" height="${snap10(opts.h)}" as="geometry"/></mxCell>`;
  return ok(`${cell}\n`, `# matched: ${slug} - ${entry.display}\n`);
}
